src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -495,7 +495,17 @@ private static object CreateInstance([DynamicallyAccessedMembers(DynamicallyAcce
             Type itemType = collectionType.GenericTypeArguments[0];
             MethodInfo addMethod = collectionType.GetMethod("Add", DeclaredOnlyLookup);
 
-            foreach (IConfigurationSection section in config.GetChildren())
+            IEnumerable<IConfigurationSection> children;
+            if (config.GetChildren().Any(a => long.TryParse(a.Key, out _)))
```

Comment:
Can we extract this into a helper method and use it in both places? Also it would be nice to write a comment or two about what this is doing, and why.